Sources/NIOPosix/SocketChannel.swift

```diff
@@ -423,12 +443,24 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
             throw err
         }
 
-        self.pendingWrites = PendingDatagramWritesManager(msgs: eventLoop.msgs,
-                                                          iovecs: eventLoop.iovecs,
-                                                          addresses: eventLoop.addresses,
-                                                          storageRefs: eventLoop.storageRefs,
-                                                          controlMessageStorage: eventLoop.controlMessageStorage)
-
+#if SWIFTNIO_USE_IO_URING && os(Linux)
+        let msgs = UnsafeMutableBufferPointer<MMsgHdr>.allocate(capacity: 1)
+        let iovecs = UnsafeMutableBufferPointer<IOVector>.allocate(capacity: 1)
+        let storageRefs = UnsafeMutableBufferPointer<Unmanaged<AnyObject>>.allocate(capacity: 1)
+        let addresses = UnsafeMutableBufferPointer<sockaddr_storage>.allocate(capacity: 1)
+        let controlMessageStorage = UnsafeControlMessageStorage.allocate(msghdrCount: 1)
+#else
+        let msgs = eventLoop.msgs
+        let iovecs = eventLoop.iovecs
+        let addresses = eventLoop.addresses
+        let storageRefs = eventLoop.storageRefs
+        let controlMessageStorage = eventLoop.controlMessageStorage
+#endif
+        self.pendingWrites = PendingDatagramWritesManager(msgs: msgs,
```

Comment:
I don't love having the pending writes managers take ownership of pointers allocated outside their scope. Can we have two different initialisers, one where they allocate the pointers themselves and one where they don't, and then keep track of what happened with a flag? Then we can drop most of the compile time conditionals, which will make our lives a lot easier.

Reply:
I do not love that either...
Had an impression having a different initialiser will require bigger changes in tests...
But may be we could use something different?
For example we could introduce a cache of IOVector buffers in the EventLoop, and then each channel could use that cache. Without URing that cache will always have a one buffer. With URing - probably more than one but less than total number of channels. We could save some memory in that case. And the buffer management logic will not require any conditional compilation at all in that case. What do you think?

Reply:
I could probably try to implement that cache in a separate branch,
then we will merge it, and then will merge into uring-out branch, and then we will continue with it.
What do you think?

Reply:
So that's https://github.com/apple/swift-nio/pull/2358 - just commenting to tie it together with this comment chain.